### PR TITLE
Fix heredoc dollar error

### PR DESCRIPTION
**Pull request by Nicholas-Lin**
## Description

This PR fixes problems with parsing heredocs. Heredocs may contain both string literals and variables. The problem was that the previous `scanner.cc` mistakenly interpreted every instance of `$` as the beginning of a variable. This is not the case as string literals may also contain `$`. This is demonstrated below:

### Previous state
```
<<<EOT
    $()                        <-- Incorrectly interpreted as invalid variable and errors out
EOT;
````

### Fixed state
```
<<<EOT
    $()                        <-- Correctly interpreted as a string literal
EOT;
````

## Test cases

I wrote 3 test cases. The first test case covers the common case (accounts for ~518 errors in WA). The other 2 test cases are edge cases I have identified that fail (accounts for 12 errors in WA). Considering the sma, it might be good to go ahead and merge and focus on more prevalent errors. 

1. `heredoc-dollar`
- This test covers the normal/expected case when there is a dollar sign string literal
- This case passes

2. `heredoc-dollar-no-lead-space`
- This tests an edge case when there is no leading space between the string literal and leading heredoc ID
- This case currently fails

3. `heredoc-dollar-var`
- This tests an edge case when there is a real variable in between string literals that contain `$`.
- This case currently fails

### CODE REVIEW

File: src/scanner.cc [hunk 1]
```diff
@@ -248,7 +248,14 @@ struct Scanner {
             next();
           } else {
             stop();
-            return true;
+            next();
+            std::regex variable("[a-zA-Z_\x80-\xff]");
```
Comment (antoniodejesusochoasolano) [inline src/scanner.cc:252]: Let's add a function that checks if a character is a valid identifier character rather than use a regex here.

The regex library doesn't seem super intuitive to use (imo), and I suspect it might be slightly less performant than a character range comparison. Performance probably isn't a big concern here but using a regex also doesn't seem to offer much value over a function with character comparisons.

Also, I think Tree-sitter actual converts the regexes used in grammar rules into character range comparisons. You'll find something like the snippet below in `src/parser.c`.

```c
case 86:
  ACCEPT_TOKEN(sym_identifier);
  if (('0' <= lookahead && lookahead <= '9') ||
      ('A' <= lookahead && lookahead <= 'Z') ||
      lookahead == '_' ||
      ('a' <= lookahead && lookahead <= 'z') ||
      (128 <= lookahead && lookahead <= 255)) ADVANCE(86);
  END_STATE();
```

Hacklang's own lexer defines valid identifier characters with [similar logic](https://github.com/facebook/hhvm/blob/fab9abb6b527ca3f3576148ef89e1ecb20bff95d/hphp/hack/src/parser/core/lexer.rs#L322-L328):

```rust
fn is_name_letter(c: char) -> bool {
    (c == '_')
        || ('0' <= c && c <= '9')
        || ('a' <= c && c <= 'z')
        || ('A' <= c && c <= 'Z')
        || ('\x7f' <= c)
}
```

File: src/scanner.cc [hunk 2]
```diff
@@ -234,7 +235,6 @@ struct Scanner {
           } else {
             stop();
             next();
-
```
Comment (antoniodejesusochoasolano) [inline src/scanner.cc:237]: nit: Doesn't seem like we need this change? Reducing drive-by whitespace changes like this make git history slightly easier to navigate.

File: test/cases/literals/heredoc-dollar-no-lead-space.hack [hunk 3]
```diff
@@ -0,0 +1,4 @@
+// This test currently fails but should pass
+<<<EOT
+$()
+EOT;
```
Comment (antoniodejesusochoasolano) [inline test/cases/literals/heredoc-dollar-no-lead-space.hack:4]: Let's add final new lines to these files... I forget why it matters (or if it actually matters, but that red stop-looking sign looks scary). You can update VSCode to automatically add a final new line after saving a file.

Screen Shot 2021-07-01 at 12 33 54 AM

File: test/cases/literals/heredoc-dollar-no-lead-space.hack [hunk 4]
```diff
@@ -0,0 +1,4 @@
+// This test currently fails but should pass
```
Comment (antoniodejesusochoasolano) [inline test/cases/literals/heredoc-dollar-no-lead-space.hack:1]: If we add the same is-identifier-char check that you added to scanner.cc t[he first time we look  for `$` in `scan_body`](https://github.com/antoniodejesusochoasolano/tree-sitter-hack/blob/81944f086839fd70d27d0396c75e080eb516490c/src/scanner.cc#L214-L217), I think that fixes this issue. Something like:

```c
if (!is_nowdoc && prefix.empty()) {
  if (peek() == '{') next();
  if (peek() == '$') {
    next();
    if (is_identifier_char(peek())) {
      return false;
    }
  }
}
```

File: test/cases/literals/heredoc-dollar-var.hack [hunk 5]
```diff
@@ -0,0 +1,4 @@
+// This test currently fails but should pass
```
Comment (antoniodejesusochoasolano) [inline test/cases/literals/heredoc-dollar-var.hack:1]: This is failing even with your changes?

File: src/scanner.cc (hunk 2 above)
Comment (Nicholas-Lin) [reply src/scanner.cc:237]: I fully agree, my bad for being messy on my end and not catching that!

File: src/scanner.cc (hunk 1 above)
Comment (Nicholas-Lin) [reply src/scanner.cc:252]: Good idea! I'll make these changes.

File: src/scanner.cc [hunk 6]
```diff
@@ -271,6 +275,14 @@ struct Scanner {
     }
   }
 
+  
+  // This function returns true if c is a valid starting character of a name/identifier
+  bool is_identifier_start_char(char c){
+    return (c == '_')
+        || ('a' <= c && c <= 'z')
+        || ('A' <= c && c <= 'Z');
```
Comment (antoniodejesusochoasolano) [inline src/scanner.cc:283]: Characters in the ranges between 128 and 255 are valid in variables apparently. Let's check those too. The generated `src/parser.c` file does this: `(128 <= lookahead && lookahead <= 255))`. We'd need to change the parameter type to `int32_t` which is the type of `lexer->lookahead`.

Can we also add a test for this in one of the here doc tests? We can use a variable with a character like `$€`.

File: src/scanner.cc [hunk 7]
```diff
@@ -271,6 +275,14 @@ struct Scanner {
     }
   }
 
+  
+  // This function returns true if c is a valid starting character of a name/identifier
+  bool is_identifier_start_char(char c){
```
Comment (antoniodejesusochoasolano) [inline src/scanner.cc:280]: I need to add an auto-formatter, so sorry about these nits: but can we change `){` to `) {` and remove the extra white space line on line 278 (there's two consecutive blank lines in 277 and 278).

File: src/scanner.cc (hunk 2 above)
Comment (antoniodejesusochoasolano) [reply src/scanner.cc]: No problem! These things feel extra nitpicky and I really should have lint rules to catch them so people can automatically catch them :)

File: src/scanner.cc (hunk 7 above)
Comment (Nicholas-Lin) [reply src/scanner.cc:280]: Good catch! For some reason I did not have the auto-format on save enabled in VSCode but it should be good after enabling it.

File: test/cases/literals/heredoc-dollar-var.hack [hunk 8]
```diff
@@ -0,0 +1,6 @@
+// This test ensures that variables within heredocs are interpreted as such.
+// ÿ is 255 in unicode which is a valid variable identifier.
```
Comment (antoniodejesusochoasolano) [inline test/cases/literals/heredoc-dollar-var.hack:2]: 🙇
